File: 1914. cyclically-rotating-a-grid.py
class Solution:
    def rotateGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:

        m = len(grid)
        n = len(grid[0])

        layers = min(m, n) // 2

        for layer in range(layers):

            nums = []

            top = layer
            bottom = m - layer - 1
            left = layer
            right = n - layer - 1

            for j in range(left, right + 1):
                nums.append(grid[top][j])

            for i in range(top + 1, bottom):
                nums.append(grid[i][right])

            for j in range(right, left - 1, -1):
                nums.append(grid[bottom][j])

            for i in range(bottom - 1, top, -1):
                nums.append(grid[i][left])

            length = len(nums)

            rotate = k % length

            rotated = nums[rotate:] + nums[:rotate]

            idx = 0

            for j in range(left, right + 1):
                grid[top][j] = rotated[idx]
                idx += 1

            for i in range(top + 1, bottom):
                grid[i][right] = rotated[idx]
                idx += 1

            for j in range(right, left - 1, -1):
                grid[bottom][j] = rotated[idx]
                idx += 1

            for i in range(bottom - 1, top, -1):
                grid[i][left] = rotated[idx]
                idx += 1

        return grid


# Test cases
if __name__ == "__main__":
    from typing import List
    from copy import deepcopy
    
    solution = Solution()
    
    # Example 1
    grid1 = [[40,10],[30,20]]
    grid1_copy = deepcopy(grid1)
    result1 = solution.rotateGrid(grid1_copy, 1)
    print(f"Example 1: Input: grid = {grid1}, k = 1")
    print(f"Output: {result1}")
    print(f"Expected: [[10,20],[40,30]]\n")
    
    # Example 2
    grid2 = [[1,2,3,4],[5,6,7,8],[9,10,11,12],[13,14,15,16]]
    grid2_copy = deepcopy(grid2)
    result2 = solution.rotateGrid(grid2_copy, 2)
    print(f"Example 2: Input: grid = {grid2}, k = 2")
    print(f"Output: {result2}")
    print(f"Expected: [[3,4,8,12],[2,11,10,16],[1,7,6,15],[5,9,13,14]]")

File: 1914. cyclically-rotating-a-grid.java
class Solution {
    public int[][] rotateGrid(int[][] grid, int k) {

        int m = grid.length;
        int n = grid[0].length;

        int layers = Math.min(m, n) / 2;

        for (int layer = 0; layer < layers; layer++) {

            ArrayList<Integer> nums = new ArrayList<>();

            int top = layer;
            int bottom = m - layer - 1;
            int left = layer;
            int right = n - layer - 1;

            for (int j = left; j <= right; j++) {
                nums.add(grid[top][j]);
            }

            for (int i = top + 1; i <= bottom - 1; i++) {
                nums.add(grid[i][right]);
            }

            for (int j = right; j >= left; j--) {
                nums.add(grid[bottom][j]);
            }

            for (int i = bottom - 1; i >= top + 1; i--) {
                nums.add(grid[i][left]);
            }

            int len = nums.size();

            int rotate = k % len;

            int[] rotated = new int[len];

            for (int i = 0; i < len; i++) {
                rotated[i] = nums.get((i + rotate) % len);
            }

            int idx = 0;

            for (int j = left; j <= right; j++) {
                grid[top][j] = rotated[idx++];
            }

            for (int i = top + 1; i <= bottom - 1; i++) {
                grid[i][right] = rotated[idx++];
            }

            for (int j = right; j >= left; j--) {
                grid[bottom][j] = rotated[idx++];
            }

            for (int i = bottom - 1; i >= top + 1; i--) {
                grid[i][left] = rotated[idx++];
            }
        }

        return grid;
    }
    
    // Helper function to print 2D array
    private static void print2DArray(int[][] arr) {
        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            System.out.print("[");
            for (int j = 0; j < arr[i].length; j++) {
                System.out.print(arr[i][j]);
                if (j < arr[i].length - 1) System.out.print(",");
            }
            System.out.print("]");
            if (i < arr.length - 1) System.out.print(",");
        }
        System.out.println("]");
    }
    
    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Example 1
        int[][] grid1 = {{40, 10}, {30, 20}};
        int[][] result1 = solution.rotateGrid(grid1, 1);
        System.out.println("Example 1: Input: grid = [[40,10],[30,20]], k = 1");
        System.out.print("Output: ");
        print2DArray(result1);
        System.out.println("Expected: [[10,20],[40,30]]\n");
        
        // Example 2
        int[][] grid2 = {{1,2,3,4},{5,6,7,8},{9,10,11,12},{13,14,15,16}};
        int[][] result2 = solution.rotateGrid(grid2, 2);
        System.out.println("Example 2: Input: grid = [[1,2,3,4],[5,6,7,8],[9,10,11,12],[13,14,15,16]], k = 2");
        System.out.print("Output: ");
        print2DArray(result2);
        System.out.println("Expected: [[3,4,8,12],[2,11,10,16],[1,7,6,15],[5,9,13,14]]");
    }
}